[사용자] 회원정보 조회 및 수정 기능 추가

awesomeo184 opened the pull request:
resolve: #13 
resolve: #19 

리뷰 반영사항
- 컨트롤러로 직접 엔티티가 넘어가는 일이 없도록 반환 타입을 Dto로 바꿨습니다.
- 컨트롤러에 노출되지 않는 경우 엔티티를 그대로 반환합니다. 대신 내부적으로 사용되는 메서드는 private으로 접근 제한자를 수정했습니다.
- ControllerAdvice를 RestControllerAdivice로 수정했습니다.

awesomeo184 committed:
refactor(Member): MemberConverter 추가, 패키지 구조 일부 수정

- MemberService와 MemberRepository를 각각 member.service, member.repository 패키지로 이동.
- MemberConverter 추가

awesomeo184 committed:
feat(OAuth2): 소셜 로그인 api 응답 로직 추가

awesomeo184 committed:
feat(Member): 검증 애너테이션 추가

awesomeo184 committed:
Merge branch 'dev' of https://github.com/prgrms-web-devcourse/Team_Price_Offer_BE into feat/#13

awesomeo184 committed:
fix(Member): URI 수정

awesomeo184 committed:
feat(Member): 이메일 중복 기능 추가

awesomeo184 committed:
feat(Member): 개인정보 조회, 수정 기능 추가

인증 처리에서 loginId를 사용하는데 소셜 로그인의 경우 카카오 고유 회원 식별 아이디가 들어간다. 테이블의 email 필드에도 카카오 고유 회원 식별 아이디가 들어간다.

hun-ca commented on src/main/java/com/prgrms/offer/domain/member/model/dto/EmailLoginRequest.java:
```diff
@@ -2,9 +2,16 @@
+
 @Getter
 public class EmailLoginRequest {
+    @NotBlank
+    @Email
     private String email;
```
@ NotBlank는 빼도 되지 않을까요

hun-ca commented on src/main/java/com/prgrms/offer/domain/member/model/dto/EmailLoginRequest.java:
```diff
@@ -2,9 +2,16 @@
+    @Email
     private String email;
+
+    @NotBlank
     private String password;
```
오늘 프론트와 논의해서 비밀번호 검증로직도 추가하면 좋을 것 같습니다!

hun-ca commented on src/main/java/com/prgrms/offer/domain/member/model/entity/Member.java:
```diff
@@ -27,7 +28,8 @@
 
-    private int appleLevel;
+    @Builder.Default
+    private int appleLevel = 1;
```
위에 필드값들은 @Column 을 안붙여도 자동으로 컬럼으로 등록되나여

awesomeo184 replied:
넵

hun-ca commented on src/main/java/com/prgrms/offer/domain/member/service/MemberService.java:
```diff
@@ -1,31 +1,41 @@
+    private final S3ImageUploader s3ImageUploader;
 
     @Transactional(readOnly = true)
     public Member login(String principal, String credentials) {
```
서비스에서 엔티티를 반환하는걸 허용하는 쪽으로 갈지 의논해볼 필요가 있을것 같아요

hun-ca commented on src/main/java/com/prgrms/offer/domain/member/service/MemberService.java:
```diff
@@ -38,6 +48,15 @@ public Member login(String principal, String credentials) {
+
     @Transactional(readOnly = true)
     public Optional<Member> findByPrincipal(String principal) {
         Assert.hasText(principal, "principal must be provided.");
```
검증을 여기서밖에 할 수 없는 구조인지 궁금합니다

hun-ca commented on src/main/java/com/prgrms/offer/domain/member/service/MemberService.java:
```diff
@@ -50,24 +69,28 @@ public Member login(String principal, String credentials) {
         return memberRepository.findByPrincipal(principal);
     }
```
예외처리 로직이 필요할것 같아요

hun-ca commented on src/main/java/com/prgrms/offer/domain/member/service/MemberService.java:
```diff
@@ -50,24 +69,28 @@ public Member login(String principal, String credentials) {
+                .nickname(request.getNickname())
                 .principal(request.getEmail())
-                .password(passwordEncoder.encode(request.getPassword()))
                 .address(request.getAddress())
-                .nickname(request.getNickname())
+                .appleLevel(1)
+                .password(passwordEncoder.encode(request.getPassword()))
                 .build();
 
```
컨버터로 빠져야할 것 같아요

hun-ca commented on src/main/java/com/prgrms/offer/domain/member/service/MemberService.java:
```diff
@@ -89,6 +112,7 @@ public Member join(OAuth2User oAuth2User, String registrationId) {
                     String profileImage = (String) properties.get("profile_image");
                     return memberRepository.save(
                             Member.builder()
```
컨버터로 빠져야 할 것 같습니다

hun-ca commented on src/main/java/com/prgrms/offer/domain/member/service/MemberService.java:
```diff
@@ -105,4 +129,32 @@ public Member join(OAuth2User oAuth2User, String registrationId) {
     public Optional<Member> findByProviderId(String kakaoId) {
         return memberRepository.findByProviderId(kakaoId);
     }
```
위에 findByProviderAndProviderId() 메소드는 없어도 될 것 같습니다

hun-ca commented on src/main/java/com/prgrms/offer/domain/member/model/dto/ProfileEdit.java:
```diff
@@ -0,0 +1,24 @@
+
+@Getter
+public class ProfileEdit {
+    @NotBlank
+    private String nickname;
+    @NotBlank
+    private String address;
+    @NotBlank
+    private String profileImageUrl;
```
DB에 길이 제한이 있으니 여기서 먼저 길이 검증을 걸어주시는게 어떨까요?

hun-ca commented on src/main/java/com/prgrms/offer/domain/member/model/dto/MemberCreateRequest.java:
```diff
@@ -2,13 +2,26 @@
+    @NotBlank
     private String address;
+
+    @NotBlank
```
NotBlank는 없어도 될 것 같습니당

hun-ca reviewed (commented):
리뷰 답변 부탁드립니다!!!

그리고 추가로 confirmedPassword와 password를 비교해주는 로직이 클라이언트에 있다면 서버에서는 confirmedPassword자체를 받을 필요가 없지 않을까요?

BeautterLife commented on src/main/java/com/prgrms/offer/core/oauth2/OAuth2AuthenticationSuccessHandler.java:
```diff
@@ -15,27 +20,44 @@
+
+        MemberResponse memberResponse = memberConverter.toMemberResponse(member, token);
+
+        ObjectMapper objectMapper = new ObjectMapper();
```
지역변수보다 클래스 멤버나 외부로 빼는건 어떤가요?

BeautterLife reviewed (commented):
리뷰 달았습니다.
수고하셨습니다!

awesomeo184 committed:
refactor(Member): 코드리뷰 내용 반영

- 컨트롤러로 직접 엔티티가 넘어가는 일이 없도록 반환 타입을 Dto로 바꿨습니다.
- 컨트롤러에 노출되지 않는 경우 엔티티를 그대로 반환합니다. 대신 내부적으로 사용되는 메서드는 private으로 접근 제한자를 수정했습니다.
- ControllerAdvice를 RestControllerAdivice로 수정했습니다.

hun-ca reviewed (approved):
고생하셨습니다~!
